# Homework 6: projections

This notebook projects a vector $y$ onto the column space of $X$ using three different procedures: ordinary projection, Gram-Schmidt orthogonalization and projection, and QR decompositon. 

Begin by importing the required libraries. 

In [1]:
import numpy as np
from numpy.linalg import inv
from numpy.linalg import qr
from scipy.linalg import orth
from numpy.linalg import norm

Next, we create the data vectors $y$ and $X$:


In [2]:
y = np.array([1,3,-3])
y.shape = 3, 1
X = np.array([[1,0], [2,0], [-6, 2]])

### Ordinary Projection

Next, create a function that computes a projection matrix $P = X(X'X)^{-1} X'$. This can then be re-used later:

In [3]:
def projection(X):
    XT = np.transpose(X)
    XTX = XT@X
    XTXinv = inv(XTX)
    P = X@XTXinv@XT
    return P

Now compute $Py$ using our data matrix, $X$:

In [4]:
P = projection(X)
Py = P@y

print(Py)

[[ 1.4]
 [ 2.8]
 [-3. ]]


### Gram-Schmidt orthogonalization and projection

Now note that the columns of $X$ are linearly independent, so we may use the Gram-Schmidt orthogonalization algorithm. The following code implements the algorithm via sequential construction of orthonomal vectors by projecting the columns of $X$ onto its orthogonal complement.  

In [ ]:
Xhi  = X.shape[0]
Xlen = X.shape[1]

M    = np.identity(Xhi)
U    = np.zeros([Xhi,Xlen])

for i in range(Xlen):
    v      = M@X[:,i]
    vnorm  = norm(v)
    U[:,i] = v/vnorm
    Xtmp = X[:,:i+1]
    P = projection(Xtmp)
    M = np.identity(Xhi) - P

Finally, compute the projection $Py = UU'y$, where $U$ is obtained from the Gram-Schmidt orthogonalization above. 

In [34]:
UT = np.transpose(U)
Py = U@UT@y

print(Py)

[[ 1.4]
 [ 2.8]
 [-3. ]]


### QR decomposition and projection

Finally, we construct the projection using the QR decomposition

In [29]:
Q, R = qr(X)
QT = np.transpose(Q)
QQT = Q@QT
Py_qr = QQT@y

print(Py_qr)

[[ 1.4]
 [ 2.8]
 [-3. ]]


Using the QR decomposition, we have exactly the same fitted value vector, $\hat{y}$, yet again. 

Thus, we see that the ordinary orthogonal projection method, the Gram-Schmidt algorithm, and the QR decomposition all produce the same orthogonal projection: $Py = [1.4, 2.8, -3.0]'$.   